### PR TITLE
Dependency collation: changes the strong dependencies of eureka、Apollo、Nacos into optional dependency. 

### PR DESCRIPTION
**Pull request by linhuaichuan**

### CODE REVIEW

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/AbstractRpcRemotingServer.java [hunk 1]
```diff
@@ -126,9 +125,9 @@ public void start() {
             .childOption(ChannelOption.TCP_NODELAY, true)
             .childOption(ChannelOption.SO_SNDBUF, nettyServerConfig.getServerSocketSendBufSize())
             .childOption(ChannelOption.SO_RCVBUF, nettyServerConfig.getServerSocketResvBufSize())
-            .childOption(ChannelOption.WRITE_BUFFER_WATER_MARK,
-                new WriteBufferWaterMark(nettyServerConfig.getWriteBufferLowWaterMark(),
-                    nettyServerConfig.getWriteBufferHighWaterMark()))
+            .childOption(ChannelOption.WRITE_BUFFER_HIGH_WATER_MARK,
+                nettyServerConfig.getWriteBufferHighWaterMark())
+            .childOption(ChannelOption.WRITE_BUFFER_LOW_WATER_MARK, nettyServerConfig.getWriteBufferLowWaterMark())
```
Comment (slievrly) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/AbstractRpcRemotingServer.java]: why？I just changed

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/mysql/keyword/MySQLKeywordChecker.java [hunk 2]
```diff
@@ -18,7 +18,7 @@
 
 import com.alibaba.fescar.rm.datasource.undo.KeywordChecker;
 
-import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang.StringUtils;
```
Comment (github-ygy) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/mysql/keyword/MySQLKeywordChecker.java]: why change the lang3 to lang?

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/AbstractRpcRemotingClient.java [hunk 3]
```diff
@@ -332,7 +331,7 @@ public void dispatch(long msgId, ChannelHandlerContext ctx, Object msg) {
         List<String> availList = new ArrayList<>();
         List<InetSocketAddress> availInetSocketAddressList = RegistryFactory.getInstance().lookup(
             transactionServiceGroup);
-        if (!CollectionUtils.isEmpty(availInetSocketAddressList)) {
+        if (null != availInetSocketAddressList && !availInetSocketAddressList.isEmpty()) {
```
Comment (github-ygy) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/AbstractRpcRemotingClient.java]: better use com.alibaba.fescar.common.util.CollectionUtils#isNotEmpty

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/RmRpcClient.java [hunk 4]
```diff
@@ -203,7 +202,7 @@ private void reconnect() {
         } catch (Exception exx) {
             LOGGER.error(exx.getMessage());
         }
-        if (CollectionUtils.isEmpty(availList)) {
+        if (availList == null || availList.isEmpty()) {
```
Comment (github-ygy) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/RmRpcClient.java]: better use com.alibaba.fescar.common.util.CollectionUtils#isNotEmpty

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/TmRpcClient.java [hunk 5]
```diff
@@ -177,7 +176,7 @@ private void reconnect() {
         } catch (Exception exx) {
             LOGGER.error(exx.getMessage());
         }
-        if (CollectionUtils.isEmpty(availList)) {
+        if (null == availList || availList.isEmpty()) {
```
Comment (github-ygy) [inline core/src/main/java/com/alibaba/fescar/core/rpc/netty/TmRpcClient.java]: better use com.alibaba.fescar.common.util.CollectionUtils#isEmpty

File: discovery/src/main/java/com/alibaba/fescar/discovery/loadbalance/AbstractLoadBalance.java [hunk 6]
```diff
@@ -30,7 +28,7 @@
 
     @Override
     public <T> T select(List<T> invokers) {
-        if (CollectionUtils.isEmpty(invokers)) {
+        if (null == invokers || invokers.isEmpty()) {
```
Comment (github-ygy) [inline discovery/src/main/java/com/alibaba/fescar/discovery/loadbalance/AbstractLoadBalance.java]: better use com.alibaba.fescar.common.util.CollectionUtils#isEmpty

File: core/src/main/java/com/alibaba/fescar/core/rpc/netty/AbstractRpcRemotingServer.java (hunk 1 above)
Comment (linhuaichuan) [reply core/src/main/java/com/alibaba/fescar/core/rpc/netty/AbstractRpcRemotingServer.java]: Sorry, it's carelessly covered.

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/mysql/keyword/MySQLKeywordChecker.java (hunk 2 above)
Comment (linhuaichuan) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/mysql/keyword/MySQLKeywordChecker.java]: common-lang3 is not  fescar-all dependency transfer

**Comment by slievrly** [comment]
pr title in English，will be written as commit log。

File: common/src/main/java/com/alibaba/fescar/common/util/CollectionUtils.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package com.alibaba.fescar.common.util;
+
+import java.util.Collection;
+
+/**
+ * The type Collection utils.
+ */
+public class CollectionUtils {
```
Comment (slievrly) [inline common/src/main/java/com/alibaba/fescar/common/util/CollectionUtils.java]: copyright author, need to implement by yourself？

**Comment by jovany-wang** [comment]
Rewrite the title in Eenglish!

File: common/src/main/java/com/alibaba/fescar/common/util/CollectionUtils.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        return null != coll && !coll.isEmpty();
+    }
+    
+    /**
+     * Is empty boolean.
+     *
+     * @param coll the coll
+     * @return the boolean
+     */
+    public static boolean isEmpty(Collection<?> coll) {
```
Comment (jovany-wang) [inline common/src/main/java/com/alibaba/fescar/common/util/CollectionUtils.java]: better to rename this to `isNullOrEmpty` to explicitly declear its name.
And more, you can put this method into common module and combine it with `isNullOrEmpty` for `String` with generic.

File: config/src/main/java/com/alibaba/fescar/config/ConfigurationFactory.java [hunk 9]
```diff
@@ -59,16 +57,12 @@ public static Configuration getInstance() {
             case Nacos:
                 try {
                     configuration = new NacosConfiguration();
-                } catch (NacosException e) {
+                } catch (Exception e) {
```
Comment (jovany-wang) [inline config/src/main/java/com/alibaba/fescar/config/ConfigurationFactory.java]: What's the purpose of this change?

Comment (linhuaichuan) [reply config/src/main/java/com/alibaba/fescar/config/ConfigurationFactory.java]: exclude nacos explicit dependencies ,

**Comment by zhangthen** [comment]
There are conflicting files

**Comment by xingfudeshi** [comment]
I have communicated with the author, and the author agreed to close this PR. I will pay attention to this issue in the subsequent PR, and I really appreciate the author's contribution.
